lib/core-integration/src/Test/Integration/Framework/DSL.hs

```diff
@@ -865,6 +873,31 @@ eventuallyUsingDelay delay io = do
         threadDelay delay
         trial
 
+-- similar as @eventually@ but returns IO a
+eventually'
+    :: IO a
+    -> IO a
+eventually' = eventuallyUsingDelay' (500 * ms)
+  where
+    ms = 1000
+
+-- similar as @eventuallyUsingDelay@ but returns IO a
+eventuallyUsingDelay'
+    :: Int
+    -> IO a
+    -> IO a
+eventuallyUsingDelay' delay io = do
+    winner <- race (threadDelay $ 60 * oneSecond) trial
+    case winner of
+        Left _ -> fail
+            "waited more than 60s for action to eventually resolve."
+        Right a ->
+            return a
+  where
+    trial = io `catch` \(_ :: SomeException) -> do
+        threadDelay delay
+        trial
```

Comment:
Since the definition of this function is virtually identical to `eventuallyUsingDelay`, we could actually rewrite `eventuallyUsingDelay` as follows:

```hs
eventuallyUsingDelay :: Int -> IO () -> IO ()
eventuallyUsingDelay = eventuallyUsingDelay'
```